resteasy-client/src/main/java/org/jboss/resteasy/client/jaxrs/internal/ClientConfigProviderFilter.java

```diff
@@ -0,0 +1,42 @@
+package org.jboss.resteasy.client.jaxrs.internal;
+
+import org.jboss.resteasy.client.jaxrs.spi.ClientConfigProvider;
+import org.jboss.resteasy.util.BasicAuthHelper;
+
+import javax.ws.rs.client.ClientRequestContext;
+import javax.ws.rs.client.ClientRequestFilter;
+import javax.ws.rs.core.HttpHeaders;
+import java.io.IOException;
+import java.net.URI;
+
+/**
+ * Client filter that will attach authorization header with either HTTP Basic auth or Bearer token auth.
+ * Credentials and token are loaded from ClientConfigProvider implementation.
+ *
+ * @author dvilkola@redhat.com
+ */
+public class ClientConfigProviderFilter implements ClientRequestFilter {
+
+    private final ClientConfigProvider clientConfigProvider;
+
+    public ClientConfigProviderFilter(final ClientConfigProvider clientConfigProvider) {
+        this.clientConfigProvider = clientConfigProvider;
+    }
+
+    @Override
+    public void filter(ClientRequestContext requestContext) throws IOException {
+        if (requestContext.getHeaderString(HttpHeaders.AUTHORIZATION) == null) {
+            URI uri = requestContext.getUri();
+            String token = clientConfigProvider.getBearerToken(uri);
+            if (token != null) {
+                requestContext.getHeaders().putSingle(HttpHeaders.AUTHORIZATION, "Bearer " + token);
+            } else {
+                String username = clientConfigProvider.getUsername(uri);
+                String password = clientConfigProvider.getPassword(uri);
+                if (username != null && password != null) {
+                    requestContext.getHeaders().putSingle(HttpHeaders.AUTHORIZATION, BasicAuthHelper.createHeader(username, password));
+                }
```

Comment:
Should it be considered an error if a ClientConfigProvider exists but it doesn't hold either a token or username/password? Or maybe there whould be a WARNING?

Reply:
@ronsigal Thanks for review! It is not considered error or warning if ClientConfigProvider does not hold token or credentials. It is not necessary to provide those, it could be that it is present only to provide SSL context.